src/ucan.js

```diff
@@ -0,0 +1,187 @@
+import { didKeyType, publicKeyBytesToDid, verify } from './did.js'
+import { base64url, utf8 } from './encoding.js'
+import {
+  serialize,
+  deserialize,
+  jwtAlgorithm,
+  isExpired,
+  isTooEarly,
+} from './utils.js'
+
+const TYPE = 'JWT'
+const VERSION = '0.8.0'
+
+export { KeyPair } from './keypair.js'
+
+/**
+ * @param {import("./types").BuildParams} params
+ */
+export async function build(params) {
+  const keypair = params.issuer
+  const didStr = publicKeyBytesToDid(keypair.publicKey)
+  const payload = buildPayload({
+    ...params,
+    issuer: didStr,
+  })
+  return sign(payload, keypair)
+}
+
+/**
+ * Build parts
+ *
+ * @param {import('./types').BuildPayload} params
+ */
+function buildPayload(params) {
+  const {
+    issuer,
+    audience,
+    capabilities = [],
+    lifetimeInSeconds = 30,
+    expiration,
+    notBefore,
+    facts,
+    proofs = [],
+    // addNonce = false,
+  } = params
+
+  // Timestamps
+  const currentTimeInSeconds = Math.floor(Date.now() / 1000)
+  const exp = expiration || currentTimeInSeconds + lifetimeInSeconds
+
+  /** @type {import('./types').UcanPayload} */
+  const payload = {
+    aud: audience,
+    att: capabilities,
+    exp,
+    fct: facts,
+    iss: issuer,
+    nbf: notBefore,
+    // nnc: addNonce ? util.generateNonce() : undefined,
+    prf: proofs,
```

Comment:
I wonder if it's worth putting a check here to ensure that the proofs are encoded? Or even automated encoding of them if they're not, such as:

```javascript
proofs = proofs.map((prf) => isUcan(prf) ? prf : serialize(prf));
```

It feels like it would be good to catch that kind of rookie error at this point, rather than have an error thrown from somewhere later/deeper in the chain when the processing happens.

Reply:
we should make an issue on this